src/zelos/emulator/base.py

```
@@ -14,14 +14,285 @@
+        page_addr = address & PageTable.PAGE_MASK
+
+        # Fast path if all data is on one page.
+        try:
+            mr = self._pages[page_addr]
+            offset = address - mr.address
+            if offset + size < mr.size:
+                return mr.host_data[offset : offset + size]
+        except (IndexError, ValueError):
+            pass
+        except KeyError:
+            raise MemoryReadUnmapped(
+                f"Read unmapped memory at address 0x{address:x}"
+            )
+
+        # Slow path if data spans multiple pages.
+        data = bytearray()
+        while size > 0:
+            try:
+                mr = self._pages[address & PageTable.PAGE_MASK]
+            except KeyError:
+                raise MemoryReadUnmapped(
+                    f"Read unmapped memory at address 0x{address:x}"
+                )
+            offset = address - mr.address
+            bytes_to_read = min(size, mr.size - offset)
+            data += mr.host_data[offset : offset + size]
+            size -= bytes_to_read
+            address += bytes_to_read
+        return data
+
+    def write(self, address: int, data: bytes) -> None:
+        """
+        Writes `data` to guest memory at `address`.
 
-from typing import Dict, Iterable, List
+        Args:
+            address: The address to read.
+            size: The size of the data to fetch in bytes.
+
+        Returns:
+            The data at the specified address.
+
+        """
+        if len(data) == 0:
+            return
+        page_addr = address & PageTable.PAGE_MASK
 
-from unicorn import UC_ARCH_ARM, UC_ARCH_MIPS, UC_ARCH_X86
+        # Fast path if all data is on one page.
+        try:
+            mr = self._pages[page_addr]
+            offset = address - mr.address
+            if offset + len(data) < mr.size:
+                memmove(mr.host_address + offset, data, len(data))
+                return
+        except (IndexError, ValueError):
+            pass
```

Should we be silently passing these errors?